### PR TITLE
TINKERPOP-1862 TinkerMessenger proper handling of Direction.BOTH

### PR DESCRIPTION
**Pull request by PBGraff**
New PR based off of #777 re-based on new head of tp32 branch. Need to update toy graph merged in #794 with scenario for this test and then update relevant tests.

### CODE REVIEW

**Comment by PBGraff** [comment]
Hi @spmallette, I'm trying to add to the kitchen sink graph for my message passing test and I want to make sure I'm doing it right. I'm doing this in Java and calling a custom `main()`:

```
Graph g1 = TinkerGraph.open();
Graph g2 = TinkerGraph.open();
Graph g3 = TinkerGraph.open();
    
try {
  // GraphSON v1.0
  g1.io(GraphSONIo.build(GraphSONVersion.V1_0)).readGraph("../tinkerpop/data/tinkerpop-sink.json");
  addData(g1);
  g1.io(GraphSONIo.build(GraphSONVersion.V1_0)).writeGraph("tinkerpop-sink.json");
  // GraphSON v2.0
  g2.io(GraphSONIo.build(GraphSONVersion.V2_0)).readGraph("../tinkerpop/data/tinkerpop-sink-v2d0.json");
  addData(g2);
  g2.io(GraphSONIo.build(GraphSONVersion.V2_0)).writeGraph("tinkerpop-sink-v2d0.json");
  // Gryo
  g3.io(GryoIo.build()).readGraph("../tinkerpop/data/tinkerpop-sink.kryo");
  addData(g3);
  g3.io(GryoIo.build()).writeGraph("tinkerpop-sink.kryo");
} catch (IOException e) {
  e.printStackTrace();
}
```

where 

```
private static void addData(Graph g) {
  Vertex a = g.addVertex(T.id, 1L, T.label, LABEL, PROPERTYIN, VTX_A);
  Vertex b = g.addVertex(T.id, 2L, T.label, LABEL, PROPERTYIN, VTX_B);
  a.addEdge(EDGE_LABEL, b);
  a.addEdge(EDGE_LABEL, a);
}
```

`V1_0` appears to be coming out right, but `V2_0` looks to be typed when the original isn't. Plus, I can't see how to read/write typed versions as opposed to un-typed. Gryo appears to be straightforward. Should I add a GraphML version?

**Comment by PBGraff** [comment]
Never mind, I think I got it.

**Comment by PBGraff** [comment]
@spmallette : I believe this is ready now if you can run the Docker test: `docker/build.sh -t -n -i`

**Comment by spmallette** [comment]
sorry - the process for adding new data to the sink is different. i didn't realize that the self-edge graph i'd already added wasn't sufficient for what you needed or i would have provided instructions. you actually needed to do this:

1. modify this to include your data: https://github.com/apache/tinkerpop/blob/master/tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerFactory.java#L147
2.  then do `cd tinkergraph-gremlin` and `mvn clean install -Dio` which will generate and update all the relevant files.

If you want I can go through that process for you when I merge it so that you don't need to fuss with this PR anymore. Not a big deal - I'm more concerned about the general content of the PR. Please let me know if you just want me to handle it that way.

**Comment by PBGraff** [comment]
I'll modify and let you know when I'm done so that I make sure it's setup as I need. For my test I need not just a self-edge but also an edge to another vertex.

**Comment by PBGraff** [comment]
Done

**Comment by PBGraff** [comment]
I'm not sure why this is failing, it's due to something with the javascript. I don't know how my changes potentially affected that.

File: tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerFactory.java [hunk 1]
```diff
@@ -149,6 +149,13 @@ public static void generateKitchenSink(final TinkerGraph graph) {
         g.addV("loops").property(T.id, 1000).property("name", "loop").as("me").
           addE("self").to("me").
           iterate();
+        final String LABEL = "message_passing_test";
+        final String EDGE_LABEL = "msg_pass_test_edge";
+        final String PROPERTY_IN = "propertyin";
+        Vertex a = graph.addVertex(T.id, 2000, T.label, LABEL, PROPERTY_IN, "a");
```
Comment (spmallette) [inline tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerFactory.java]: all of these sample graphs need a "name" property of some sort. that's why the build is currently failing for me.  please "final" the two `Vertex` variables.

**Comment by PBGraff** [comment]
OK, I've made that update and now this build passes.

**Comment by spmallette** [comment]
I'm sorry to say that spark integration tests are failing for your added tests. Not sure what's wrong. I made an initial fix that solved the first body of errors. I updated `ToyGraphInputRDD` to include:

```java
else if (configuration.getString(Constants.GREMLIN_HADOOP_INPUT_LOCATION).contains("sink"))
            vertices = IteratorUtils.list(IteratorUtils.map(TinkerFactory.createKitchenSink().vertices(), VertexWritable::new));
```

But the tests still fail with:

```text
[ERROR] testMessagePassingOut(org.apache.tinkerpop.gremlin.process.computer.GraphComputerTest)  Time elapsed: 5.693 s  <<< ERROR!
java.util.concurrent.ExecutionException: java.lang.IllegalStateException: Name is null
	at org.apache.tinkerpop.gremlin.process.computer.GraphComputerTest.runMPTest(GraphComputerTest.java:2741)
	at org.apache.tinkerpop.gremlin.process.computer.GraphComputerTest.testMessagePassingOut(GraphComputerTest.java:2710)
Caused by: java.lang.IllegalStateException: Name is null
Caused by: java.lang.NullPointerException: Name is null

[ERROR] testMessagePassingBoth(org.apache.tinkerpop.gremlin.process.computer.GraphComputerTest)  Time elapsed: 5.397 s  <<< ERROR!
java.util.concurrent.ExecutionException: java.lang.IllegalStateException: Name is null
	at org.apache.tinkerpop.gremlin.process.computer.GraphComputerTest.runMPTest(GraphComputerTest.java:2741)
	at org.apache.tinkerpop.gremlin.process.computer.GraphComputerTest.testMessagePassingBoth(GraphComputerTest.java:2726)
Caused by: java.lang.IllegalStateException: Name is null
Caused by: java.lang.NullPointerException: Name is null

[ERROR] testMessagePassingIn(org.apache.tinkerpop.gremlin.process.computer.GraphComputerTest)  Time elapsed: 5.198 s  <<< ERROR!
java.util.concurrent.ExecutionException: java.lang.IllegalStateException: Name is null
	at org.apache.tinkerpop.gremlin.process.computer.GraphComputerTest.runMPTest(GraphComputerTest.java:2741)
	at org.apache.tinkerpop.gremlin.process.computer.GraphComputerTest.testMessagePassingIn(GraphComputerTest.java:2694)
Caused by: java.lang.IllegalStateException: Name is null
Caused by: java.lang.NullPointerException: Name is null
```

Not sure I know what is causing that....I don't what "Name" is referring to here. I was guessing the property you just added to "sink" but it's not the same case, so......

**Comment by spmallette** [comment]
I just pushed the fix for the `ToyGraphInputRdd`:

https://github.com/apache/tinkerpop/commit/a61dd58eea353637d74086426e4258de5927a8d9

If your rebase you'll get that little fix.

**Pull request closed without merging**